### DISCUSSION TITLE
Jinx not finding installed libraries on NixOS

### DISCUSSION THREAD

**Discussion by thibautbenjamin** [minad/jinx#173] [Q&A] [answered]
Hi, first of all thank you for the great work, I have been using and very much enjoying jinx for quite some time.

However, I have recently started using NixOS, and I cannot get jinx working on there. I don't really know what's wrong, as it seems to work for other people, but for some reason, during startup, I get the following message:
```
Jinx: No dictionaries available for "en_GB fr_FR"
```

I am using the `jinx` package provided by nix,  I have the dictionaries installed in nix, the recommended way, adding the following lines to my configuration:
```
  environment.systemPackages = with pkgs; [
    (aspellWithDicts
      (dicts: with dicts; [ fr en en-computers en-science ]))
  ];
```

I could also check, installing the `enchant` package, that the dictionaries were detected:
```
❯ enchant-lsmod-2 -list-dicts
en (aspell)
en-variant_0 (aspell)
en-variant_1 (aspell)
en-variant_2 (aspell)
en-w_accents (aspell)
en-wo_accents (aspell)
en_AU (aspell)
en_AU-variant_0 (aspell)
en_AU-variant_1 (aspell)
en_AU-w_accents (aspell)
en_AU-wo_accents (aspell)
en_CA (aspell)
en_CA-variant_0 (aspell)
en_CA-variant_1 (aspell)
en_CA-w_accents (aspell)
en_CA-wo_accents (aspell)
en_GB (aspell)
en_GB-ise (aspell)
en_GB-ise-w_accents (aspell)
en_GB-ise-wo_accents (aspell)
en_GB-ize (aspell)
en_GB-ize-w_accents (aspell)
en_GB-ize-wo_accents (aspell)
en_GB-variant_0 (aspell)
en_GB-variant_1 (aspell)
en_GB-w_accents (aspell)
en_GB-wo_accents (aspell)
en_US (aspell)
en_US-variant_0 (aspell)
en_US-variant_1 (aspell)
en_US-w_accents (aspell)
en_US-wo_accents (aspell)
fr (aspell)
fr-40 (aspell)
fr-60 (aspell)
fr-80 (aspell)
fr-lrg (aspell)
fr-med (aspell)
fr-sml (aspell)
fr_CH (aspell)
fr_CH-40 (aspell)
fr_CH-60 (aspell)
fr_CH-80 (aspell)
fr_CH-lrg (aspell)
fr_CH-med (aspell)
fr_CH-sml (aspell)
fr_FR (aspell)
fr_FR-40 (aspell)
fr_FR-60 (aspell)
fr_FR-80 (aspell)
fr_FR-lrg (aspell)
fr_FR-med (aspell)
fr_FR-sml (aspell)
he (hspell)
he_IL (hspell)
```

However, it seems that jinx does not see these dictionaries, as when I evaluate `(mapcar #'jinx--mod-dict (ensure-list "en_US"))`, I get `(nil)`.

The package seems to be working for everyone else, so I must be doing something wrong, but I don't know what. 

Any help for debugging the situation would be greatly appreciated!

**Comment by fbergroth** [comment]
I was facing the same issue. Enchant is usingt the aspell package directly and not the wrapper created by `aspellWithDicts`.

I fixed it by just setting:
```
home.sessionVariables.ASPELL_CONF = "dict-dir ${pkgs.aspellWithDicts (ds: with ds; [en sv])}/lib/aspell";
```

**Reply by thibautbenjamin** [reply]
Hi, thanks for your help, this works when lauching from terminal, but not when launching from the app launcher

**Comment by alinnow** [comment, marked as answer]
I fixed it by not using aspell; nuspell seems to work much better without any configuration, probably hunspell would be the same (they both use hunspell dictionaries). I didn't even need to use `nuspellWithDicts`. This is important (for me) because I was running into the issue where opening Emacs from the desktop environment (either Plasma6 or macOS) wouldn't load dictionaries, but if I launched it from a shell, it would (due to the `ASPELL_CONF` variable)

The [Enchant docs](https://abiword.github.io/enchant/) make it sound like nuspell and hunspell are more modern than aspell.

Add to `home.packages` or  `environment.systemPackages`:
```nix
with pkgs.hunspellDicts; [
  # whatever
]
```
Dictionary packages are in [<nixpkgs/pkgs/development/libraries/hunspell/dictionaries.nix>](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/libraries/hunspell/dictionaries.nix)

Currently I have this in an overlay:
```nix
  enchant =
    let
      file = builtins.fetchurl {
        url = "https://raw.githubusercontent.com/alanpearce/nixpkgs/7ee75d6aa9b088922b47b69c1912b2afe000ae52/pkgs/development/libraries/enchant/2.x.nix";
        sha256 = "1y7h3fvp7ghcfb5v3h6riiv7z2bzgnxm0p696bbhfb0gwwv9q8aw";
      };
    in
    super.callPackage file {
      withHspell = false;
      withAspell = false;

      inherit (super.darwin.apple_sdk.frameworks) Cocoa;
    };
```

This will also enable AppleSpell on Darwin.

After https://github.com/NixOS/nixpkgs/pull/310613 lands ([track PR](https://nixpk.gs/pr-tracker.html?pr=310613)), this overlay can be reduced to `super.enchant.override { withAspell = false; withHspell = false; }`

**Reply by thibautbenjamin** [reply]
Hi, thanks a lot,

I am fairly new to nixOS, so I am not exactly sure how to use the overlays yet. For now I am using the session variable solution since it works well when launching emacs through the terminal. I might investigate using nuspell later, to have it working when launching emacs from an app launcher.

I'll mark your solution as an answer because it seems more complete, even though I have not tested it.

**Comment by edgar-vincent** [comment]
This `home-manager` configuration: 
```nix
  systemd.user.settings.Manager.DefaultEnvironment = {
    ASPELL_CONF = "dict-dir ${
      pkgs.aspellWithDicts (
        ds: with ds; [
          en
          en-computers
          es
          fr
        ]
      )
    }/lib/aspell";
  };
```
seems to work here, both for GUI and CLI launches of Emacs.

**Comment by larstvei** [comment]
I also struggled with enchant not finding the dictionaries installed by aspell/hunspell/nuspell using nix. The strangest part is that it does find some of the AppleSpell dictionaries, but not the Norwegian (which is what I have on my system).

I ended up adding [`(hunspellWithDicts [ hunspellDicts.nb_NO ])`](https://github.com/larstvei/nix-config/blob/d818c16e8f518acde0f1351ec893c3322c50b8ae/modules/home.nix#L5) to my packages, and adding [the following](https://github.com/larstvei/nix-config/blob/d818c16e8f518acde0f1351ec893c3322c50b8ae/modules/home.nix#L65) in home-manager:

```
  home.file.".config/enchant/hunspell/".source = "${pkgs.hunspellDicts.nb_NO}/share/hunspell/";
```

My ideal solution would be to package it directly in my Emacs flake: https://github.com/larstvei/emacs-flake/blob/main/flake.nix

But I haven't found a way to do this yet.

**Reply by goofansu** [reply]
@larstvei Thank you! It's simplest method I can find.

**Comment by RomeoV** [comment]
I had the same issue, using `enchant` with `nuspell` and `hunspell` dictionaries. Interestingly, I was able to find the dictionaries running

```bash
$ enchant-lsmod-2 -list-dicts
de_AT (nuspell)
de_BE (nuspell)
de_CH (nuspell)
de_DE (nuspell)
de_LI (nuspell)
de_LU (nuspell)
en_US (nuspell)
en_US-large (nuspell)
```
from my regular terminal, but not from `emacs`! I.e., even when running the same command from a terminal from within emacs, `enchant` would not find the dictionaries.

Ultimately, I found that the culprit was **`XDG_DATA_DIRS`**, which was set from within emacs, but not my regular terminal, and somehow setting it made `enchant` not find any dictionaries.
This is the solution I have come up with

```nix
  config = let
    hunspellDicts = with pkgs.hunspellDicts; [ en_US de_DE ];
    nuspellWithUserDicts = pkgs.nuspellWithDicts hunspellDicts;
    baseEnchant = pkgs.enchant.override {
      withAspell = false;
      withHspell = false;
      withHunspell = false;
      withNuspell = true;
    };
    # We have to circumvent an annoying problem:
    # Generally, we can "bake in" the dictionaries into =enchant= like this.
    # However, emacs sets the =xdg_data_dirs= env var, and this trips up enchant somehow!
    # Probably a but. So we have to unset it, and then everything works fine.
    # See here for more info: https://github.com/minad/jinx/discussions/173
    wrappedEnchant = pkgs.runCommand "enchant-wrapped" {
      nativeBuildInputs = [ pkgs.makeWrapper ];
      buildInputs = [ baseEnchant nuspellWithUserDicts ];
    } ''
      mkdir -p $out/bin
      makeWrapper ${baseEnchant}/bin/enchant-2 $out/bin/enchant-2 \
        --unset XDG_DATA_DIRS
      makeWrapper ${baseEnchant}/bin/enchant-lsmod-2 $out/bin/enchant-lsmod-2 \
        --unset XDG_DATA_DIRS
      ln -s ${baseEnchant}/lib $out/lib
      ln -s ${baseEnchant}/share $out/share
    '';

  in {
    environment = {
      systemPackages = (with pkgs; [
        # ...
        wrappedEnchant
      ])
  };
```
